slapo/schedule.py

```diff
@@ -311,24 +386,53 @@ def sync(self, mode="backward"):
         Case 3: (shard x, shard_in w) -> partial sum -> allreduce
                 -> (replica x, shard_out w).
             In this case, backward does not need allreduce, so mode should be 'forward'.
+        Case 4: (shard x, shard_in w) -> partial sum -> reduce-scatter
+                -> ... -> allgather -> full output.
+            This case breaks the allreduce in case 3 to reduce-scatter and allgather,
+            which is called "sequential parallelism". In this case, we also need
+            to specify the allgather point in kwargs. For example,
+            sch["out_proj"].sync(mode="forward_defer_gather", gather_at=sch)
         """
         assert (
             "output_type" in self.metadata.shard
         ), "output_type is missing in {mod}.schedule_metadata.shard"
         output_type = self.metadata.shard["output_type"]
 
-        if mode in {"forward", "both"}:
+        if mode in {"forward", "forward_defer_gather", "both"}:
             if output_type == "partition":
+                if mode == "forward_defer_gather":
+                    raise ValueError(f"mode {mode} is not supported for {output_type}")
+
                 # Case 1
                 gather_axis = self.metadata.shard["gather_axis"]
                 sync_fn = partial(
                     all_gather_forward_output, dim=gather_axis, group=self.group
                 )
             elif output_type == "partial":
-                # Case 3
-                def sync_fn(output):
-                    dist.all_reduce(output, op=dist.ReduceOp.SUM, group=self.group)
-                    return output
+                if mode == "forward_defer_gather":
+                    if "gather_at" not in kwargs:
+                        raise ValueError(
+                            "gather_at is missing in kwargs for mode "
+                            "forward_defer_gather"
+                        )
+
+                    # Case 4
+                    gather_sch, gather_axis = kwargs["gather_at"]
+                    sync_fn = partial(
+                        reduce_scatter_forward_output, dim=gather_axis, group=self.group
+                    )
```

Comment:
Do we need to verify whether the gather_axis is valid?

Reply:
It is verified in `reduce_scatter_forward_output`. The only requirement is the dimension being scattered has to be divisible by TP size.